File: code/js/139.js
/**
 * @param {string} s
 * @param {string[]} wordDict
 * @return {boolean}
 */
var wordBreak = function (s, wordDict) {
  const dp = (s, wordDict, memo = {}) => {
    if (s in memo) return memo[s]
    if (s === "") return true

    for (let word of wordDict) {
      if (s.startsWith(word)) {
        // replace the first occurrence
        const y = s.replace(word, '')
        if (dp(y, wordDict, memo) === true) {
          memo[s] = true
          return memo[s]
        }
      }
      else {
        // if match failed, skip the recursion
        continue
      }
    }

    memo[s] = false
    return memo[s]
  }
  return dp(s, wordDict)

  const bruteForce = (s, wordDict) => {
    if (s === "") return true

    for (let word of wordDict) {
      if (s.startsWith(word)) {
        // replace the first occurrence
        const y = s.replace(word, '')
        if (dp(y, wordDict) === true) {
          return true
        }
      }
      else {
        // if match failed, skip the recursion
        continue
      }
    }

    return false
  }

};

console.log(wordBreak("leetcode", ["leet", "code"]) == true)
console.log(wordBreak("applepenapple", ["apple", "pen"]) == true)
console.log(wordBreak("catsandog", ["cats", "dog", "sand", "and", "cat"]) == false)
console.log(wordBreak("ccbb", ["bc", "cb"]) == false)
console.log(wordBreak("bccbbc", ["bc", "cb"]) == true)

File: code/java/139.java
import java.util.HashMap;
import java.util.List;
import java.util.Map;

class Solution {
    public boolean wordBreak(String s, List<String> wordDict) {
        Map<String, Boolean> memo = new HashMap<>();
        return dp(s, wordDict, memo);
    }

    private boolean dp(String s, List<String> wordDict, Map<String, Boolean> memo) {
        if (memo.containsKey(s))
            return memo.get(s);
        if (s.equals(""))
            return true;

        for (String word : wordDict) {
            String y;
            if (s.startsWith(word)) {
                y = s.substring(word.length());
            } else {
                continue;
            }
            if (dp(y, wordDict, memo) == true) {
                memo.put(s, true);
                return memo.get(s);
            }
        }

        memo.put(s, false);
        return memo.get(s);
    }

    public boolean bruteForce(String s, List<String> wordDict) {
        if (s.equals(""))
            return true;

        for (String word : wordDict) {
            String y;
            if (s.startsWith(word)) {
                y = s.substring(word.length());
            } else {
                continue;
            }
            if (bruteForce(y, wordDict) == true) {
                return true;
            }
        }

        return false;
    }

    public static void main(String[] args) {
        Solution solution = new Solution();
        System.out.println(solution.wordBreak("leetcode", List.of("leet", "code")) == true);
        System.out.println(solution.wordBreak("applepenapple", List.of("apple", "pen")) == true);
        System.out.println(solution.wordBreak("catsandog", List.of("cats", "dog", "sand", "and", "cat")) == false);
        System.out.println(solution.wordBreak("ccbb", List.of("bc", "cb")) == false);
        System.out.println(solution.wordBreak("cars", List.of("car", "ca", "rs")) == true);
    }
}